File: python/p002/solve.py
"""Problem 002 - Even Fibonacci Numbers"""

from collections.abc import Iterator

def fib(n: int) -> Iterator[int]:
  i, j = 1, 2
  yield 1
  while j < n:
    yield j
    i, j = j, i+j

if __name__ == "__main__":
  print(sum(fibi for fibi in fib(4000000) if fibi % 2 == 0))

File: golang/p002/solve.go
package p002

// Problem 2 - Even Fibonacci Numbers

func SumEvenFibonacciUntil(n int64) int64 {
	sum := int64(0)
	for fibn := range fibonacci(n) {
		if fibn&1 == 0 {
			sum += fibn
		}
	}
	return sum
}

func fibonacci(limit int64) <-chan int64 {
	fibs := make(chan int64)
	go func() {
		var i, j int64 = 1, 2
		fibs <- 1
		defer close(fibs)

		for j < limit {
			fibs <- j
			i, j = j, i+j
		}
	}()
	return fibs
}
